text/Issue 8 - Plug-in based architecture.md

```diff
@@ -0,0 +1,265 @@
+role of a `Message` is to standardize (de)serialization.
+
+A crate `standardized` provides a library of standard `Message`s to cover most uses:
+- boolean;
+- string;
+- various number types;
+- dates;
+- time;
+- temperature;
+- image;
+- `RawJSON`;
+- `RawBinary`;
+- ...
+
+Use of a `Message` is decided per-channel, by the Adapter developer.
+
+When implementing a channel, adapter developers are encouraged to pick one
+of the full-featured `Message`s (e.g. temperature or time), or can also provide
+new implementations of `Message`, if they prefer. Additional instances of `Message`
+can easily be published by third-party developers on third-party crates.
+
+While this is less optimal, Adapter developers can fallback to `RawJSON` or `RawBinary`
+if they prefer implementing their own mechanism internally, although this may
+come at some cost performance cost (typically, they may need to reparse the
+same value several times from JSON or Binary).
+
+
+### Implementation
+
+```rust
+trait Message: Sized {
+  /// Developer-oriented human-readable name, used mainly for error messages.
+  ///
+  /// For instance, `bool`, `positive number`, `temperature`.
+  fn name() -> String;
+
+  /// Values carried by messages of this type.
+  type Type: Any;
+
+  /// A mechanism used to serialize values of this type.
+  type Serializer: Serializer<Source = Self::Type>;
+
+  /// A mechanism used to deserialize values of this type.
+  type Deserializer: Deserializer<Target = Self::Type>;
+}
+
+/// A mechanism used to serialize values of a given type.
+trait Serializer {
+  type Source;
+
+  /// Serialize a value of type `Source` to JSON. If the value has binary
+  /// components, they should be placed in `binary`.
+  fn serialize(source: &Self::Source, binary: &mut BinaryComponents) -> Result<JSON, SerializeError>;
+}
+trait Deserializer {
+  type Target;
+
+  /// Deserialize a value of type `Target`. If the value has binary components,
+  /// they can be fetched from `binary`.
+  fn deserialize(source: &JSON, binary: &BinaryComponents) -> Result<Self::Target, ParseError>;
+}
+```
+
+Note that `BinaryComponents` are taken from ongoing work on https://github.com/fxbox/taxonomy/issues/37
+and are not specific to this RFC.
+
+
+## Channel features
+
+We replace `ChannelKind` with `Feature`. As `ChannelKind`, the role of a
+`Feature` is to allow an application to discover channels (respectively service)
+that implement a given feature (respectively feature set), and to standardize
+the (de)serialization format for this channel.
+
+A crate `standardized` provides a library of standard `Feature`s to cover a number
+of uses:
+
+- garage door opened;
+- lock unlocked;
+- oven temperature;
+- ...
+
+Use of a `Feature` is decided per-channel, by the Adapter developer.
+
+Third-party developers are encouraged to pick a standardized `Feature`, but
+are not forced to. Additional instances of `Feature` can easily be published
+by third-party developers to third-party crates.
+
```

Comment:
Why does `Feature` need to be generic over `Message`? In what way is this better than having channels broadcast a simple string-array of "features"?